data migrations: move consumer offsets using API calls

bashtanov opened the pull request:
As @dotnwat suggested, I've split the PR, this is the preparations part:
1. Make c/leader_router work as a sharded service, and also support non-copiable request type
2. moving consumer-offset-related types to different files
3. k/s/group_manager: functions for bulk get/set consumer offsets, and a bugfix

Migrations subsystem changes coming in the next PR

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes
* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

bashtanov committed:
c/leader_router: rename shutdown() to stop()

for to use the class directly as a sharded service

bashtanov committed:
cluster: move offsets_snapshot.h from cloud_metadata to the upper level

It is going to be used for data migrations. While data migrations
themselves require cloud storage, offsets are being migrated via API
completely bypassing cloud storage.

bashtanov committed:
k/s/group_manager: add private function to dump selected groups

to be used for group migrations

bashtanov committed:
c/offsets_snapshot: get/set offsets RPC types

bashtanov committed:
k/s/data_migration_group_proxy: get/set consumer offsets

bashtanov committed:
k/s/group_manager: snapshot groups for migrating away

bashtanov committed:
k/s/group_manager: move group_offsets_snapshot_result into the class

It is only used in this class, and does not belong to offsets_snapshot.h
since it was moved to cluster/

bashtanov committed:
k/s/group_manager: refactor to allow merge behaviour

bashtanov committed:
k/s/group_manager: correct error code on partition not found

Previously an implicit cast between enums resulted in returning
`kafka::error_code::request_timed_out`.

bashtanov committed:
k/s/group_manager: function to set group offsets for migrations

Unlike the one used for recovery, this one is idempotent.

dotnwat reviewed (commented):
Taking a quick peek at the commits, it seems like you could probably split this PR up into 3 smaller ones, and likely accelerate the review process.

1) One PR will tangential or othogonal changes to offset migration. For example, this commit: 

> container/chunked_vector: allow constructing from non-sized ranges

looks like it was authored over a month ago. It probably could have been in a PR the day it was written.

2) Core feature work

3) Optionally, changes from category (1) that are difficult to add into the first PR because of the commit history.

dotnwat commented on src/v/kafka/server/group_manager.cc:
```diff
@@ -888,6 +888,11 @@ group_manager::do_snapshot_groups(
+    return do_bulk_write_offsets(std::move(snap), false);
+}
+
+ss::future<kafka::error_code>
+group_manager::do_bulk_write_offsets(group_offsets_snapshot snap, bool merge) {
```
nit: the declaration for this is in the previous commit. either it should move into this commit, or the definition should move into the previous commit.

bashtanov replied:
Thanks, will fix. I do check bisectability for large PRs, so it does not affect it.

dotnwat replied:
> i don't think the declaration/definition split affects bisectability

yeh, since there aren't any calls to it. otherwise i think we'd get a missing symbol or something during linking.

dotnwat commented on src/v/kafka/server/group_manager.cc:
```diff
@@ -891,6 +891,33 @@ group_manager::recover_offsets(group_offsets_snapshot snap) {
 
+ss::future<cluster::set_group_offsets_reply>
+group_manager::set_group_offsets(cluster::set_group_offsets_request req) {
+    return do_bulk_write_offsets(std::move(req.group_offsets), false)
```
what makes it idempotent? the most substantive piece of this function looks like it is `do_bulk_write_offsets`, but it uses `merge = false` which was the previous behavior?

dotnwat reviewed (commented):
lgtm. i don't think the declaration/definition split affects bisectability, so no need to fix that. i think the question about idempotency is the only real question.

also, thanks for splitting up the previous PR!

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/71257-->
<!--0cc6274c-b8cc-43e5-8f1d-b62d6f7c8bc1-->
<!--title-->
## Retry command for [Build#71257](https://buildkite.com/redpanda/redpanda/builds/71257)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/transactions/tx_coordinator_migration_test.py::TxCoordinatorMigrationTest.test_migrating_tx_manager_coordinator@{"upgrade":false,"with_failures":true}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--cf3b8054-03a3-4c65-922b-481fd8cdcd70-->
<!--TEST RESULTS-->
<!--583e97e8-7f6f-4c43-bb62-851ef935c5a6-->
<!--TEST RESULTS-->
<!--0bf7df93-5df4-4011-9fa2-711cf79f57d5-->
<!--TEST RESULTS-->
<!--3a5ea86a-4f3d-4b35-857d-38f25d2e4f46-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/71257">build#71257</a></summary>


| test_class                     | test_method                           | test_arguments                            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                |
|:-------------------------------|:--------------------------------------|:------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DataMigrationsMultiClusterTest | test_basic                            | null                                      | integration | https://buildkite.com/redpanda/redpanda/builds/71257#0198d336-922f-45f3-9286-99aa5681bd4a | FLAKY         | 19/21    | upstream reliability is '96.39468690702088'. current run reliability is '90.47619047619048'. drift is 5.9185 and the allowed drift is set to 50. The test should PASS |
| TxCoordinatorMigrationTest     | test_migrating_tx_manager_coordinator | {"upgrade": false, "with_failures": true} | integration | https://buildkite.com/redpanda/redpanda/builds/71257#0198d336-922c-4f6e-b226-da5eaf4a2441 | FLAKY         | 8/21     | upstream reliability is '100.0'. current run reliability is '38.095238095238095'. drift is 61.90476 and the allowed drift is set to 50. The test should FAIL          |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/71442">build#71442</a></summary>


| test_class                 | test_method                           | test_arguments                            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                       |
|:---------------------------|:--------------------------------------|:------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------|
| TxCoordinatorMigrationTest | test_migrating_tx_manager_coordinator | {"upgrade": false, "with_failures": true} | integration | https://buildkite.com/redpanda/redpanda/builds/71442#0198eafb-f6be-49c3-9841-3f402caf87db | FLAKY         | 5/21     | upstream reliability is '100.0'. current run reliability is '23.809523809523807'. drift is 76.19048 and the allowed drift is set to 50. The test should FAIL |
| TxCoordinatorMigrationTest | test_migrating_tx_manager_coordinator | {"upgrade": true, "with_failures": true}  | integration | https://buildkite.com/redpanda/redpanda/builds/71442#0198eafb-f6bf-4d44-b878-2ac617fccaaa | FLAKY         | 11/21    | upstream reliability is '100.0'. current run reliability is '52.38095238095239'. drift is 47.61905 and the allowed drift is set to 50. The test should PASS  |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/71532">build#71532</a></summary>


| test_class                 | test_method                           | test_arguments                            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                     |
|:---------------------------|:--------------------------------------|:------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-------------------------------------------------------------------------------------------------------------------------------------------|
| MasterTestSuite            | test_replica_pair_frequency           |                                           | unit        | https://buildkite.com/redpanda/redpanda/builds/71532#0198f163-c7b3-4017-a93d-81152ae1c8b7 | FAIL          | 0/1      |                                                                                                                                            |
| TxCoordinatorMigrationTest | test_migrating_tx_manager_coordinator | {"upgrade": false, "with_failures": true} | integration | https://buildkite.com/redpanda/redpanda/builds/71532#0198f18c-d3ef-44fb-9901-9085474d1ea5 | FLAKY         | 31/70    | upstream reliability is '100.0'. current run reliability is '25.0'. drift is 75.0 and the allowed drift is set to 50. The test should FAIL |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/71653">build#71653</a></summary>


| test_class               | test_method          | test_arguments                                                                                                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                     |
|:-------------------------|:---------------------|:---------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 2, "compaction_mode": "adjacent_merge", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/71653#0198f83f-fb75-412d-839a-953f19322593 | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

joe-redpanda commented on src/v/kafka/server/group_manager.cc:
```diff
@@ -839,11 +900,43 @@ ss::future<group_offsets_snapshot_result> group_manager::snapshot_groups(
         co_await ss::maybe_yield();
     }
-    co_return snapshots;
+    co_return std::move(snapshots);
```
move needed?

dotnwat replied:
yeh good question. do we get NRVO with future<T> return values?

bashtanov replied:
NRVO initializes the to-be-returned variable directly in the function call’s result object. It cannot apply to `co_return`, which works as a call to `promise.return_value` function under the hood.

It can be optimised as a [move instead of copy](https://stackoverflow.com/a/72513585), but
~1) it's not guaranteed (NRVO isn't either BTW)~
2) explicit `std::move` call won't make any harm

Finally, there's 45 existing occurrences of `co_return std::move(an_identifier);` in the repo. If one day we find `std::move` call redundant or harmful in this situation we can bulk fix them in a separate PR.

nvartolomei replied:
> it's not guaranteed

Intuitively I would assume that in this case it is actually a requirement to make the operand an rvalue. Otherwise, the the code could fail to compile under different compilers implementations for a move only type. Is that ok? Not versed in interpreting C++ standard lingua.

nvartolomei replied:
Btw, not that rvalue cast is implicitly applied only for local variables, parameters and couple of other cases. There are genuine cases where you require/want a std::move. I.e.`co_return std::move(this.foo)` if foo is move-only type.

bashtanov replied:
Thanks @nvartolomei, your comments made me reread the standard link and check my assumptions too.

If the expression used in `co_return` is NOT a (possibly parenthesized) id-expression it's well described in your comment above: move can only be explicit, should we need/want it.

If it is an id-expression (or, simply said, an identifier) the standard says "overload resolution to select ... the `return_value` overload to call is first performed as if the expression or operand were an rvalue". It sounds a bit vague, as once it is found it is not clear what are the semantics of the call: whether move operator is to be called at all.

Well,
1) [Godbolt](https://godbolt.org/z/oao4E3hda) shows that whether move ctor is called only depends on whether `return_value` accepts a value or an rvalue reference, not whether we wrap `co_return` argument into an `std::move(...)` (one can play with commented lines to make sure).
2) The standard section referred above describes `co_return` argument as an "implicitly movable entity", so it's reasonable to assume the call is meant to perform an implicit move.
3) As usual, if move ctor has no side effects the compiler may optimize it out, whether the move was explicit or implicit.

So, yes it is guaranteed. And `std::move` call makes no difference.

joe-redpanda reviewed (dismissed):
one nitpick, otherwise looks good

mmaslankaprv commented on src/v/kafka/server/group_manager.cc:
```diff
@@ -888,6 +888,11 @@ group_manager::do_snapshot_groups(
+
+ss::future<kafka::error_code>
+group_manager::do_bulk_write_offsets(group_offsets_snapshot snap, bool merge) {
     vassert(!snap.groups.empty(), "Group data must not be empty");
```
nit: log entries in this method seems to be outdated now

bashtanov replied:
Do you mean the word "recover" does not relate to ingesting and storing CG offset data as part of an inbound migration? What about "restore", do you think it's still OK?

mmaslankaprv commented:
one nit, lgtm otherwise

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/71442-->
<!--c16b2703-976a-43ce-9c1d-4a5b70cbe0f0-->
<!--title-->
## Retry command for [Build#71442](https://buildkite.com/redpanda/redpanda/builds/71442)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/transactions/tx_coordinator_migration_test.py::TxCoordinatorMigrationTest.test_migrating_tx_manager_coordinator@{"upgrade":false,"with_failures":true}
```
<!--/body-->

bashtanov commented:
/ci-repeat 1
skip-unit
dt-repeat=50
dt-nodes=10
tests/rptest/transactions/tx_coordinator_migration_test.py::TxCoordinatorMigrationTest.test_migrating_tx_manager_coordinator@{"upgrade":false,"with_failures":true}

bashtanov committed:
c/offsets_snapshot: implement copy()

bashtanov commented:
I have removed the attempt for using non-copiable types as RPC requests or replies. It was incorrect and does not look easily doable, as we may need copies for retrying. So I implemented copy() for these types instead. @mmaslankaprv could you re-review please?
